nimbus/src/dbcache.rs

```diff
@@ -0,0 +1,89 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::enrollment::get_enrollments;
+use crate::error::{Error, Result};
+use crate::persistence::Database;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::sync::Mutex;
+
+// This module manages an in-memory cache of the database, so that some
+// functions exposed by nimbus can return results without blocking on any
+// IO. Consumers are expected to call our public `update()` function whenever
+// the database might have changed.
+
+// This struct is the cached data. This is never mutated, but instead
+// recreated every time the cache is updated.
+struct CachedData {
+    pub experiment_branches: HashMap<String, String>,
```

Comment:
The more I look at it, the more I think we should be able to read `get_active_experiments` from any thread.

Reply:
> The more I look at it, the more I think we should be able to read `get_active_experiments` from any thread.

What's the use-case?

Reply:
And just to clarify - you *can* call it from any thread :) The semantics here are really more about "is it exposed as a synchronous function?" and "does it perform IO?". Platforms will choose to not call such functions from the main thread.

While that seems like splitting hairs, I think the distinction is worthwhile. So in practice, a synchronous, IO-performing function can be called by any thread other than the main thread. If it is exposed as an async function, it can be called just fine from the main thread (even though that implies part of the framework dispatches it to a background thread.)

This is analogous to JS on Desktop - this function will be callable from the main thread, but it will return a promise. I doubt there will be a requirement on Desktop that this be callable synchronously from the main thread. I'm interested to know what use-cases on Fenix would make it a requirement there?

Reply:
I think this is related to the [Blocking] / [Non-Blocking] annotation discussed elsewhere, so I've filed mozilla/uniffi-rs#378.

Reply:
Hmm. Yes I could've been clearer.

Currently, we're building UI tools to exercise Nimbus: `getActiveExperiments()` and `getExperimentBranches()` may benefit from being called from the main thread without triggering [StrictMode][1].

A pattern we could arrive at is:

1. read synchronously, without using I/O so we can use it from the main thread; 
2. write everything asynchronously.

At the moment, `getExperimentBranch()` is callable without I/O, but `getExperimentBranches()` is not.

The language specific framework on top of `NimbusClient` can make the API more ergonomic by dispatching writes on to worker threads from the main thread. It cannot make the API nicer if it needs to read from the main thread but Rust insists on doing I/O.

[1]: https://develozper.android.com/reference/android/os/StrictMode